File: sdks/typescript/packages/token/src/verifier.ts
/**
 * @author Sudomimus Contributors
 * @package Token
 * @namespace Verifier
 * @description Token verifier
 */

import { ACCESS_TOKEN_TYPE, REFRESH_TOKEN_TYPE } from "./constants.js";
import type {
    AccessToken,
    PublicKeyResolver,
    RefreshToken,
    TokenVerifierOptions,
} from "./declare.js";
import { TokenError } from "./error.js";
import {
    parseAccessToken,
    parseRefreshToken,
    peekTokenBody,
    peekTokenHeader,
} from "./parse.js";

export class TokenVerifier {

    private readonly _resolver: PublicKeyResolver;

    public constructor(options: TokenVerifierOptions) {

        this._resolver = options.resolver;
    }

    public async verifyAccessToken(jwt: string): Promise<AccessToken> {

        return this._verify(jwt, ACCESS_TOKEN_TYPE, parseAccessToken) as Promise<AccessToken>;
    }

    public async verifyRefreshToken(jwt: string): Promise<RefreshToken> {

        return this._verify(jwt, REFRESH_TOKEN_TYPE, parseRefreshToken) as Promise<RefreshToken>;
    }

    private async _verify(
        jwt: string,
        expectedTokenType: string,
        parser: (jwt: string) => AccessToken | RefreshToken | null,
    ): Promise<AccessToken | RefreshToken> {

        const header = peekTokenHeader(jwt);
        const body = peekTokenBody(jwt);

        if (header === null || body === null) {

            throw new TokenError("INVALID_JWT", "Token is not a parseable JWT.");
        }

        if (header.typ !== expectedTokenType) {

            throw new TokenError(
                "WRONG_TOKEN_TYPE",
                `Expected token type "${expectedTokenType}", got "${String(header.typ ?? "")}".`,
            );
        }

        const audience = body.aud;

        if (typeof audience !== "string" || audience.length === 0) {

            throw new TokenError(
                "MISSING_AUDIENCE",
                "Token is missing the `aud` (applicationAnchor) payload claim.",
            );
        }

        const keyId = header.kid;

        if (typeof keyId !== "string" || keyId.length === 0) {

            throw new TokenError(
                "MISSING_KEY_ID",
                "Token is missing the `kid` signing-key identifier.",
            );
        }

        const parsed: AccessToken | RefreshToken | null = parser(jwt);

        if (parsed === null) {

            throw new TokenError("INVALID_JWT", "Token claims do not match the 4.0.0 contract.");
        }

        if (!parsed.verifyExpiration(new Date())) {

            throw new TokenError("EXPIRED", "Token has expired.");
        }

        const publicKey: string = await this._resolver(audience, keyId);

        if (!parsed.verifySignature(publicKey)) {

            throw new TokenError(
                "INVALID_SIGNATURE",
                "Token signature does not match the application public key.",
            );
        }

        return parsed;
    }
}

File: sdks/python/packages/sudomimus-token/src/sudomimus_token/verifier.py
"""End-to-end token verification (structure, media type, audience, expiry, signature)."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from datetime import UTC, datetime
from typing import TypeVar

from .errors import TokenError, TokenErrorCode
from .models import AccessTokenBody, AccessTokenHeader, RefreshTokenBody, RefreshTokenHeader
from .parser import parse_access_token, parse_refresh_token, peek_body, peek_header
from .token import JwtToken

ACCESS_TOKEN_TYPE = "vnd.sudomimus.application-access+jwt"
REFRESH_TOKEN_TYPE = "vnd.sudomimus.application-refresh+jwt"

PublicKeyResolver = Callable[[str, str], str]
AsyncPublicKeyResolver = Callable[[str, str], Awaitable[str]]

_HeaderT = TypeVar("_HeaderT", AccessTokenHeader, RefreshTokenHeader)
_BodyT = TypeVar("_BodyT", AccessTokenBody, RefreshTokenBody)


def _now_utc() -> datetime:
    return datetime.now(tz=UTC)


def _check_before_signature(
    jwt: str,
    expected_token_type: str,
    parser: Callable[[str], JwtToken[_HeaderT, _BodyT]],
    now: datetime,
) -> tuple[JwtToken[_HeaderT, _BodyT], str, str]:
    """Run every check that does not need the public key; return token + audience.

    Peek the header first so a wrong-type token surfaces as
    ``WRONG_TOKEN_TYPE`` rather than ``INVALID_JWT``.
    """
    peeked = peek_header(jwt)
    if peeked.typ != expected_token_type:
        raise TokenError(
            TokenErrorCode.WRONG_TOKEN_TYPE,
            f'Expected token type "{expected_token_type}", got "{peeked.typ or ""}".',
        )

    payload = peek_body(jwt)
    audience = payload.get("aud")
    if not isinstance(audience, str) or not audience:
        raise TokenError(
            TokenErrorCode.MISSING_AUDIENCE,
            "Token is missing the `aud` (applicationAnchor) payload claim.",
        )

    key_id = peeked.kid
    if not key_id:
        raise TokenError(
            TokenErrorCode.MISSING_KEY_ID,
            "Token is missing the `kid` signing-key identifier.",
        )

    parsed = parser(jwt)
    if not parsed.verify_expiration(now):
        raise TokenError(TokenErrorCode.EXPIRED, "Token has expired.")

    return parsed, audience, key_id


def _check_signature(
    parsed: JwtToken[_HeaderT, _BodyT], public_key_pem: str
) -> JwtToken[_HeaderT, _BodyT]:
    if not parsed.verify_signature(public_key_pem):
        raise TokenError(
            TokenErrorCode.INVALID_SIGNATURE,
            "Token signature does not match the application public key.",
        )
    return parsed


class TokenVerifier:
    """Synchronous verifier driven by a public-key resolver."""

    def __init__(
        self,
        resolver: PublicKeyResolver,
        *,
        clock: Callable[[], datetime] = _now_utc,
    ) -> None:
        self._resolver = resolver
        self._clock = clock

    def verify_access_token(self, jwt: str) -> JwtToken[AccessTokenHeader, AccessTokenBody]:
        return self._verify(jwt, ACCESS_TOKEN_TYPE, parse_access_token)

    def verify_refresh_token(self, jwt: str) -> JwtToken[RefreshTokenHeader, RefreshTokenBody]:
        return self._verify(jwt, REFRESH_TOKEN_TYPE, parse_refresh_token)

    def _verify(
        self,
        jwt: str,
        expected_token_type: str,
        parser: Callable[[str], JwtToken[_HeaderT, _BodyT]],
    ) -> JwtToken[_HeaderT, _BodyT]:
        parsed, audience, key_id = _check_before_signature(
            jwt, expected_token_type, parser, self._clock()
        )
        return _check_signature(parsed, self._resolver(audience, key_id))


class AsyncTokenVerifier:
    """Asynchronous verifier with an awaitable public-key resolver."""

    def __init__(
        self,
        resolver: AsyncPublicKeyResolver,
        *,
        clock: Callable[[], datetime] = _now_utc,
    ) -> None:
        self._resolver = resolver
        self._clock = clock

    async def verify_access_token(
        self, jwt: str
    ) -> JwtToken[AccessTokenHeader, AccessTokenBody]:
        return await self._verify(jwt, ACCESS_TOKEN_TYPE, parse_access_token)

    async def verify_refresh_token(
        self, jwt: str
    ) -> JwtToken[RefreshTokenHeader, RefreshTokenBody]:
        return await self._verify(jwt, REFRESH_TOKEN_TYPE, parse_refresh_token)

    async def _verify(
        self,
        jwt: str,
        expected_token_type: str,
        parser: Callable[[str], JwtToken[_HeaderT, _BodyT]],
    ) -> JwtToken[_HeaderT, _BodyT]:
        parsed, audience, key_id = _check_before_signature(
            jwt, expected_token_type, parser, self._clock()
        )
        return _check_signature(parsed, await self._resolver(audience, key_id))
